core/src/main/java/org/apache/iceberg/view/ViewVersionReplace.java

```diff
@@ -0,0 +1,132 @@
+import static org.apache.iceberg.TableProperties.COMMIT_NUM_RETRIES_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_TOTAL_RETRY_TIME_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_TOTAL_RETRY_TIME_MS_DEFAULT;
+
+import java.util.List;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.PropertyUtil;
+import org.apache.iceberg.util.Tasks;
+
+class ViewVersionReplace implements ReplaceViewVersion {
+  private final ViewOperations ops;
+  private final List<ViewRepresentation> viewRepresentationsToAdd = Lists.newArrayList();
+  private ViewMetadata base;
+  private Namespace defaultNamespace;
+  private String defaultCatalog;
+  private Schema schema;
+
+  ViewVersionReplace(ViewOperations ops) {
+    this.ops = ops;
+    this.base = ops.current();
+  }
+
+  @Override
+  public ViewVersion apply() {
+    Preconditions.checkState(
+        !viewRepresentationsToAdd.isEmpty(), "Cannot replace view without specifying a query");
+    Preconditions.checkState(null != schema, "Cannot replace view without specifying schema");
+
+    this.base = ops.refresh();
+
+    ViewVersion viewVersion = base.currentVersion();
+    int maxVersionId =
+        base.versions().stream()
+            .map(ViewVersion::versionId)
+            .max(Integer::compareTo)
+            .orElseGet(viewVersion::versionId);
+
+    return ImmutableViewVersion.builder()
+        .versionId(maxVersionId + 1)
+        .timestampMillis(System.currentTimeMillis())
+        .schemaId(schema.schemaId())
+        .defaultNamespace(defaultNamespace)
+        .defaultCatalog(defaultCatalog)
+        .putSummary("operation", "replace")
+        .addAllRepresentations(viewRepresentationsToAdd)
+        .build();
+  }
+
+  @Override
+  public void commit() {
+    Tasks.foreach(ops)
+        .retry(
+            PropertyUtil.propertyAsInt(
+                base.properties(), COMMIT_NUM_RETRIES, COMMIT_NUM_RETRIES_DEFAULT))
+        .exponentialBackoff(
+            PropertyUtil.propertyAsInt(
+                base.properties(), COMMIT_MIN_RETRY_WAIT_MS, COMMIT_MIN_RETRY_WAIT_MS_DEFAULT),
+            PropertyUtil.propertyAsInt(
+                base.properties(), COMMIT_MAX_RETRY_WAIT_MS, COMMIT_MAX_RETRY_WAIT_MS_DEFAULT),
+            PropertyUtil.propertyAsInt(
+                base.properties(), COMMIT_TOTAL_RETRY_TIME_MS, COMMIT_TOTAL_RETRY_TIME_MS_DEFAULT),
+            2.0 /* exponential */)
+        .onlyRetryOn(CommitFailedException.class)
+        .run(
+            taskOps -> {
+              ViewVersion newVersion = apply();
+              // nothing to do if the version didn't change
+              if (this.base.currentVersion().equals(newVersion)) {
```

Comment:
Style: no need for `this.` when not assigning to the field.

Reply:
I think there is a problem with using `equals` with Immutables.

Checking whether the version is equivalent to other versions is done in `reuseOrCreateNewViewVersionId` in the `ViewMetadata` builder. This code doesn't use the `ViewMetadata` builder, so we know that the version ID of the new version will be assigned in `apply` so this will never succeed.

In addition, it looks like the logic in `ViewMetadata.Builder.reuseOrCreateNewViewVersionId` also uses `equals`, which checks `versionId` equality. This also needs to be fixed (along with the `build` method), but doesn't need to block this PR.

For this PR, I think the replace code should construct a new `ViewMetadata` and then check to see whether the `ViewMetadata` has changed (which is already done by `ViewOperations`). That way this delegates to the logic in `reuseOrCreateNewViewVersionId`.

Reply:
thanks for pointing that out. I'll revisit the equality comparison in `reuseOrCreateNewViewVersionId()` independently from this PR.

Also I've removed `if (this.base.currentVersion().equals(newVersion))` after refactoring the code in `ViewVersionReplace` to have an `internalApply` that constructs a `ViewMetadata` object. That way (as you pointed out), we can rely on the change-detection in `ViewOperations`

Reply:
I handled the view version ID re-assignment and de-duplication in https://github.com/apache/iceberg/pull/8664